runtime/src/crowdfund.rs

```diff
@@ -0,0 +1,539 @@
+	type OrphanedFunds: OnUnbalanced<NegativeImbalanceOf<Self>>;
+}
+
+pub type FundIndex = u32;
+
+#[derive(Encode, Decode, Clone, PartialEq, Eq)]
+#[cfg_attr(feature = "std", derive(Debug))]
+pub struct FundInfo<AccountId, Balance, Hash, BlockNumber, ParaId> {
+	/// The parachain that this fund has funded, if there is one. As long as this is `Some`, then
+	/// the funds may not be withdrawn and the fund cannot be dissolved.
+	parachain: Option<ParaId>,
+	/// The owning account who placed the deposit.
+	owner: AccountId,
+	/// The amount of deposit placed.
+	deposit: Balance,
+	/// The total amount raised.
+	raised: Balance,
+	/// Block number after which the funding must have succeeded. If not successful at this number
+	/// the everyone may withdraw their funds.
+	end: BlockNumber,
+	/// A hard-cap on the amount that may be contributed.
+	cap: Balance,
+	/// The most recent block that this had a contribution. Determines if we make a bid or not.
+	/// If this is `None`, then the last contribution was made outside of the ending period.
+	last_contribution: Option<BlockNumber>,
+	/// First slot in range to bid on; it's actually a LeasePeriod, but that's the same type as
+	/// BlockNumber.
+	first_slot: BlockNumber,
+	/// Last slot in range to bid on; it's actually a LeasePeriod, but that's the same type as
+	/// BlockNumber.
+	last_slot: BlockNumber,
+	/// The deployment data associated with this fund, if any. Once set it may not be reset. First
+	/// is the code hash, second is the initial head data.
+	deploy_data: Option<(Hash, Vec<u8>)>,
+}
+
+decl_storage! {
+	trait Store for Module<T: Trait> as Example {
+		/// Info on all of the funds.
+		Funds get(funds):
+			map FundIndex => Option<FundInfo<T::AccountId, BalanceOf<T>, T::Hash, T::BlockNumber, ParaIdOf<T>>>;
+
+		/// The total number of funds that have so far been allocated.
+		FundCount get(fund_count): FundIndex;
+
+		/// The funds that have had additional contributions during the last block. This is used
+		/// in order to determine which funds should submit updated bids.
+		NewRaise: Vec<FundIndex>;
+
+		/// True if the fund was ending at the last block.
+		WasEnding: bool;
+	}
+}
+
+decl_event!(
+	pub enum Event<T>
+		where
+	<T as system::Trait>::AccountId,
+	Balance = BalanceOf<T>,
+	ParaId = ParaIdOf<T>,
+	{
+		Created(FundIndex),
+		Contributed(AccountId, FundIndex, Balance),
+		Withdrew(AccountId, FundIndex, Balance),
+		Retiring(FundIndex),
+		Dissolved(FundIndex),
+		DeployDataFixed(FundIndex),
+		Onboarded(FundIndex, ParaId),
+	}
+);
+
+decl_module! {
+	pub struct Module<T: Trait> for enum Call where origin: T::Origin {
+		fn deposit_event<T>() = default;
+		
+		/// Create a new crowdfunding campaign for a parachain slot deposit.
+		#[weight = TransactionWeight::Basic(100_000, 10)]
+		fn create(
+			origin,
+			#[compact] end: T::BlockNumber,
+			#[compact] cap: BalanceOf<T>,
+			#[compact] first_slot: T::BlockNumber,
+			#[compact] last_slot: T::BlockNumber
```

Comment:
Missing check that `last_slot` is great then `first_slot`.

This is important to make `handle_bid` infallible.

Reply:
Could be resolved by simply calling `new_bounded` somewhere and checking it returns ok

Reply:
> Missing check that last_slot is great then first_slot.

sure

> This is important to make handle_bid infallible

`handle_bid` doesn't need to be infallible.